### PR TITLE
Document guidelines for new WASI proposals

### PR DESCRIPTION
**Pull request by alexcrichton**
I've attempted to summarize our thoughts from last week's meeting here
to jot down the suggested guidelines for adding a new WASI proposal to
this repository. I've also filled it out a bit with the two proposals we
support at this time.

### CODE REVIEW

**Review by pchickey — APPROVED** [review]

File: docs/stability-wasi-proposals-support.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+Wasmtime strives to support WASI proposals in-tree to assist with
+maintainability and provide easier to support to users for easier feedback. In
+doing so, however, we have a few guidelines around adding support for a WASI
+proposal to Wasmtime:
+
+* The proposal must be tested and exercised on CI. These tests don't need to be
+  overly exhaustive but a "smoke test" should pass. Additionally CI should not
+  be overly long compared to existing tests run, and it should also strive to be
+  relataively reliable. Wasmtime contributors may disable CI for proposals at
```
Comment (tschneidereit) [inline docs/stability-wasi-proposals-support.md:25]: Nit: "relataively". However, I think we should not soften this too much, and "strive to" already makes it a bit less strong, so perhaps just remove this?

File: docs/stability-wasi-proposals-support.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  `snapshot` directory is supported by the [`wasmtime-wasi`
+  crate](https://docs.rs/wasmtime-wasi/0.21.0/wasmtime_wasi/)
+
+* [The wasi-nn proposal](https://github.com/WebAssembly/wasi-nn) - the
+  `ephemeral` directory with the wasi-nn APIs.
+
+Wasmtime strives to support WASI proposals in-tree to assist with
+maintainability and provide easier to support to users for easier feedback. In
+doing so, however, we have a few guidelines around adding support for a WASI
+proposal to Wasmtime:
```
Comment (tschneidereit) [inline docs/stability-wasi-proposals-support.md:20]: Would it make sense to require these to have an RFC filed for them? I think that'd be good practice, and now that we're starting to actually use RFCs, it seems like there's not too much blocking that.

File: docs/stability-wasi-proposals-support.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  any time if needed, and maintainers for the proposal will be notified.
+
+* The size of the code in the wasmtime repository should not be unduly large.
+  This helps keep the size of the repo within reasonable limits and relatively
+  easy for others not working on the proposal to still be productive.
+
+* It should be possible to compile wasmtime without support for the proposal.
+  It's ok to compile it in by default (but this may not always be the right
+  choice), but Wasmtime should always be able to compile without the proposal.
+  Additionally the proposal doesn't need to support all platforms in the get-go.
```
Comment (tschneidereit) [inline docs/stability-wasi-proposals-support.md:35]: We should also have guidelines for whether features should be enabled by default at runtime, or require a flag to be set.

Maybe we can differentiate some of this by which stage in the process the proposal is in? Something like
- stage 0 proposals must be disabled by default at compile time, and must require a flag at runtime
- stage 1 proposals can be enabled by default, but must require a flag at runtime
- stage 2+ proposals can be enabled by default at compile and runtime

Comment (tschneidereit) [reply docs/stability-wasi-proposals-support.md:35]: An additional guideline might be about which namespace the API should be in. Perhaps stage 2+ can use the WASI namespacing system, but everything before some explicitly more experimental namespace? @sunfishcode and @pchickey, I'd be curious what your take is on this?

**Review by tschneidereit — COMMENTED** [review]
Thank you for writing this, looks great to me!

I left some comments with suggestions for additional guidelines that seem like fairly easy additions. If they turn out to be less simple, we probably shouldn't block on all of them though.

@mingqiusun, @abrown, @jedisct1, @3miliano since you're all working on proposals that have Wasmtime implementations, I'd be interested in hearing your feedback on these guidelines.

**Comment by jedisct1** [comment]
These guidelines look great!

But is the idea to keep full implementations of proposals within the `wasmtime` tree (as it seems to be the case for `wasi-nn`), or just `wasmtime`-specific glue to more generic code that could be shared with other runtimes?

Nobody's using `wasi-common` beyond `wasmtime`. This is sad. Each runtime wrote their own implementation instead, even runtimes written in Rust.
Maybe something we don't want to repeat for other proposals; it is not going to scale well as the number of proposals keeps growing. Having full, and `wasmtime`-specific implementations of WASI proposals here may not be optimal if we want them to get adoption in the WebAssembly world at large.

So, maybe it would be worth clarifying what exactly would be maintained in the `wasmtime` repository?

**Comment by tschneidereit** [comment]
> But is the idea to keep full implementations of proposals within the `wasmtime` tree (as it seems to be the case for `wasi-nn`), or just `wasmtime`-specific glue to more generic code that could be shared with other runtimes?

That is the idea, yes. Or at least it's one way to organize things, not the only possible one. But in the interest of full disclosure, I've argued for doing things this way, because I think there's a lot of benefit to it. In particular, we should make it as easy as possible to experiment with proposals, ideally by enabling them by default in the binaries we provide, so content authors don't have to configure and compile their own runtime. That is made significantly harder by keeping the proposals external, to the point that I don't think we should take on that burden.

> Maybe something we don't want to repeat for other proposals; it is not going to scale well as the number of proposals keeps growing. Having full, and `wasmtime`-specific implementations of WASI proposals here may not be optimal if we want them to get adoption in the WebAssembly world at large.

I agree that there's a cost to it, yes. However, there's also a very strong upside, that in my opinion wins out. The standardization process for WebAssembly with good reason requires multiple independent implementations of all features before stabilizing them in the spec. There are multiple reasons for this, and which ones are crucial depends on the nature of the proposal in question. For WASI, I think most of the time the most important one is demonstrating the viability of multiple independent implementations. I know that both in the Wasm context and for JS, this requirement has caught fundamental problems with proposals in the past. And I know that not enforcing it has caused features to be standardized that then turned out not to be implementable. In JS, one of those still exists in the spec and is still shipped by a single vendor.

File: docs/stability-wasi-proposals-support.md (hunk 2 above)
Comment (abrown) [reply docs/stability-wasi-proposals-support.md:20]: 👍 I discussed wasi-nn multiple times on Zulip and the Wasmtime meetings; an RFC would have hopefully recorded all of that in one place.

**Comment by abrown** [comment]
> content authors don't have to configure and compile their own runtime

I am in favor of @tschneidereit's point of making it really easy for users to test new WASI proposals. And I am also not opposed to having a separate repo for the sake of re-use. In wasi-nn's case, re-use is complicated by 1) the fact that the wasi-nn implementation is highly dependent on wiggle and wasmtime-wiggle, 2) Wasmtime provides wiggle/wasmtime-wiggle yet would depend on wasi-nn (a cycle), and 3) that no one is asking to re-use wasi-nn 😄.

**Review by abrown — APPROVED** [review]

File: docs/stability-wasi-proposals-support.md (hunk 3 above)
Comment (pchickey) [reply docs/stability-wasi-proposals-support.md:35]: Lets leave the namespace up to wherever the proposal is with in the WASI subgroup, but also permit an experimental namespace to be used by all proposals.

**Comment by tschneidereit** [comment]
> And I am also not opposed to having a separate repo for the sake of re-use.

To clarify, I most definitely don't want to say that we should somehow actively oppose separate repos and discourage re-use. While there are risks attached to that, ultimately the decisions around those need to be managed by proposal champions, and aren't something we can somehow address in general guidelines on the Wasmtime side.

> In wasi-nn's case, re-use is complicated by 1) the fact that the wasi-nn implementation is highly dependent on wiggle and wasmtime-wiggle, 2) Wasmtime provides wiggle/wasmtime-wiggle yet would depend on wasi-nn (a cycle), and 3) that no one is asking to re-use wasi-nn 😄.

This gets more to the heart of the issue ☺️ If we want to encourage use of our tooling around WASI APIs — which seems highly desirable, because it's good tooling 😛 — then these issues seem to apply, and are quite hard to resolve. Until someone goes and does the probably pretty hard work to get them resolved, I think we should assume them as part of the context, and structure our guidelines accordingly.

File: docs/stability-wasi-proposals-support.md (hunk 3 above)
Comment (tschneidereit) [reply docs/stability-wasi-proposals-support.md:35]: Good point that this is a WASI subgroup topic more than anything else. So perhaps a more general statement could be included here around following the WASI subgroup's rules around exposure of in-progress proposals?

**Comment by alexcrichton** [comment]
Ok I've pushed up some stronger wording, and RFC requirement, and explicitly stating what's possible depending on the phase of the proposal.

**Comment by alexcrichton** [comment]
This is quite out of date at this point, I'll push this up in the future if I get around to it with a rebase.

**Pull request closed without merging**